AnalyzingAndProfilingTools/GameOfLife/GameOfLife/AdWindow.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Windows;
+using System.Windows.Input;
+using System.Windows.Media;
+using System.Windows.Media.Imaging;
+using System.Windows.Threading;
+
+namespace GameOfLife
+{
+    class AdWindow : Window
+    {
+        private readonly DispatcherTimer adTimer;
+        private int imgNmb;     // the number of the image currently shown
+        private string link;    // the URL where the currently shown ad leads to
+        
+    
+        public AdWindow(Window owner)
+        {
+            Random rnd = new Random();
+            Owner = owner;
+            Width = 350;
+            Height = 100;
+            ResizeMode = ResizeMode.NoResize;
+            WindowStyle = WindowStyle.ToolWindow;
+            Title = "Support us by clicking the ads";
+            Cursor = Cursors.Hand;
+            ShowActivated = false;
+            MouseDown += OnClick;
+            
+            imgNmb = rnd.Next(1, 3);
+            ChangeAds(this, new EventArgs());
+
+            // Run the timer that changes the ad's image 
+            adTimer = new DispatcherTimer();
+            adTimer.Interval = TimeSpan.FromSeconds(3);
+            adTimer.Tick += ChangeAds;
+            adTimer.Start();
+        }
+
+        private void OnClick(object sender, MouseButtonEventArgs e)
+        {
+            System.Diagnostics.Process.Start(link);
+            Close();
+        }
+        
+        protected override void OnClosed(EventArgs e)
+        {
+            //Unsubscribe();

```

Comment:
Unsubscribe? https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/events/how-to-subscribe-to-and-unsubscribe-from-events#unsubscribing

Reply:
Please check update comments commit